File: codec_cortex/scalars.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Scalar value model and lexer for CORTEX 0.1.
Sections 1-2 of the independent reviewer's implementation.
"""

import re
import unicodedata
from dataclasses import dataclass
from typing import Any, List, Tuple


# ---------------------------------------------------------------------------
# 0. Utilities
# ---------------------------------------------------------------------------

def to_nfc(s: str) -> str:
    return unicodedata.normalize("NFC", s)


def utf8_bytes(s: str) -> bytes:
    return s.encode("utf-8")


# ---------------------------------------------------------------------------
# 1. Scalar value model
# ---------------------------------------------------------------------------

@dataclass
class Scalar:
    """A CORTEX scalar value with its lexeme (canonical or source)."""
    kind: str            # "string" | "atom" | "integer" | "decimal" | "boolean" | "null" | "list"
    value: Any           # str for atom/string/integer/decimal; bool; None; list[Scalar] for list
    lexeme: str          # canonical lexeme to emit

    def clone(self) -> "Scalar":
        if self.kind == "list":
            return Scalar("list", [x.clone() for x in self.value], self.lexeme)
        return Scalar(self.kind, self.value, self.lexeme)


# ---------------------------------------------------------------------------
# 2. Lexer / scalar parser
# ---------------------------------------------------------------------------

class ParseError(Exception):
    def __init__(self, code: str, message: str, line: int = 0, col: int = 0):
        self.code = code
        self.message = message
        self.line = line
        self.col = col
        super().__init__(f"{code} @ {line}:{col} — {message}")


# Atom grammar per spec §17.2 and ABNF:
#   atom = ["$" positive-integer ":"] (ALPHA / "_") *(ALPHA / DIGIT / "_" / "." / "/" / ":" / "@" / "+" / "%" / "$" / "-")
ATOM_BODY = r"[_A-Za-z][_A-Za-z0-9./:@+%$-]*"
ATOM_RE = re.compile(r"^(?:\$[0-9]+:)?" + ATOM_BODY + r"$")


_INT_RE = re.compile(r"^-?(0|[1-9][0-9]*)$")
_DEC_RE = re.compile(r"^-?(0|[1-9][0-9]*)\.[0-9]+$")


def is_atom_lexeme(s: str) -> bool:
    """Check if a string could be a valid bare atom lexeme (no quotes)."""
    if not s:
        return False
    # No whitespace, no structural delimiters
    if re.search(r"[\s\[\]{}\s,\"|]", s):
        return False
    # Length check (max 32 chars per spec)
    if len(s) > 32:
        return False
    # Must start with letter or _ (after optional $N: prefix)
    return bool(ATOM_RE.match(s))


# String escape / unescape
ESCAPE_MAP = {
    '"': '"',
    "\\": "\\",
    "n": "\n",
    "r": "\r",
    "t": "\t",
    "b": "\b",
    "f": "\f",
}
REV_ESCAPE = {
    '"': '\\"',
    "\\": "\\\\",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
    "\b": "\\b",
    "\f": "\\f",
}


def parse_string_literal(s: str) -> str:
    """Parse a quoted string literal body (without surrounding quotes)."""
    out = []
    i = 0
    while i < len(s):
        c = s[i]
        if c == "\\":
            if i + 1 >= len(s):
                raise ParseError("L005_INVALID_STRING", "Trailing backslash in string")
            nxt = s[i + 1]
            if nxt in ESCAPE_MAP:
                out.append(ESCAPE_MAP[nxt])
                i += 2
            elif nxt == "u":
                hexs = s[i + 2:i + 6]
                if len(hexs) != 4 or not all(ch in "0123456789abcdefABCDEF" for ch in hexs):
                    raise ParseError("L005_INVALID_STRING", "Bad \\u escape")
                out.append(chr(int(hexs, 16)))
                i += 6
            elif nxt == "/":
                # allow \/ as /
                out.append("/")
                i += 2
            else:
                raise ParseError("L005_INVALID_STRING", f"Unknown escape \\{nxt}")
        elif c == '"':
            raise ParseError("L005_INVALID_STRING", "Unescaped quote in string body")
        else:
            out.append(c)
            i += 1
    return "".join(out)


def emit_string_literal(value: str) -> str:
    """Emit a quoted string literal body using canonical escapes."""
    out = []
    for ch in value:
        if ch in REV_ESCAPE:
            out.append(REV_ESCAPE[ch])
        elif ord(ch) < 0x20:
            out.append("\\u%04X" % ord(ch))
        elif ord(ch) == 0x7F:
            out.append("\\u007F")
        else:
            out.append(ch)
    return '"' + "".join(out) + '"'


# ---------------------------------------------------------------------------
# 3. Attrs payload parser
# ---------------------------------------------------------------------------

class StringCursor:
    """Simple cursor over a string with position tracking."""
    def __init__(self, s: str, line: int = 1, col: int = 1):
        self.s = s
        self.i = 0
        self.line = line
        self.col = col

    def eof(self) -> bool:
        return self.i >= len(self.s)

    def peek(self, off: int = 0) -> str:
        j = self.i + off
        return self.s[j] if 0 <= j < len(self.s) else ""

    def next(self) -> str:
        c = self.s[self.i]
        self.i += 1
        if c == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return c


def parse_scalar(cur: StringCursor, in_list: bool = False) -> Scalar:
    """Parse a single scalar from cursor. Stops at , } ] | or EOF."""
    skip_inline_ws(cur)
    c = cur.peek()
    if c == '"':
        return parse_string_scalar(cur)
    elif c == "[":
        return parse_list_scalar(cur)
    else:
        return parse_atom_or_number(cur)


def skip_inline_ws(cur: StringCursor):
    while cur.peek() in " \t":
        cur.next()


def parse_string_scalar(cur: StringCursor) -> Scalar:
    assert cur.peek() == '"'
    cur.next()  # consume opening "
    body = []
    while True:
        c = cur.peek()
        if c == "":
            raise ParseError("L005_INVALID_STRING", "Unterminated string", cur.line, cur.col)
        if c == '"':
            cur.next()
            break
        if c == "\\":
            body.append(c)
            cur.next()
            nxt = cur.peek()
            if nxt == "":
                raise ParseError("L005_INVALID_STRING", "Trailing backslash", cur.line, cur.col)
            body.append(nxt)
            cur.next()
            if nxt == "u":
                for _ in range(4):
                    h = cur.peek()
                    if h == "":
                        raise ParseError("L005_INVALID_STRING", "Bad \\u escape", cur.line, cur.col)
                    body.append(h)
                    cur.next()
        else:
            body.append(c)
            cur.next()
    raw = "".join(body)
    value = parse_string_literal(raw)
    lex = emit_string_literal(value)
    return Scalar("string", value, lex)


def parse_list_scalar(cur: StringCursor) -> Scalar:
    assert cur.peek() == "["
    cur.next()  # consume [
    items = []
    skip_inline_ws(cur)
    if cur.peek() == "]":
        cur.next()
        return Scalar("list", items, "[]")
    while True:
        v = parse_scalar(cur, in_list=True)
        items.append(v)
        skip_inline_ws(cur)
        c = cur.peek()
        if c == ",":
            cur.next()
            skip_inline_ws(cur)
            continue
        elif c == "]":
            cur.next()
            break
        else:
            raise ParseError("L007_INVALID_LIST", f"Expected , or ] got {c!r}", cur.line, cur.col)
    lex = "[" + ",".join(it.lexeme for it in items) + "]"
    return Scalar("list", items, lex)


def parse_atom_or_number(cur: StringCursor) -> Scalar:
    """Parse a bare atom/integer/decimal/boolean/null token."""
    start = cur.i
    while True:
        c = cur.peek()
        if c == "" or c in " \t\r\n,}]|":
            break
        cur.next()
    raw = cur.s[start:cur.i]
    if raw == "true":
        return Scalar("boolean", True, "true")
    if raw == "false":
        return Scalar("boolean", False, "false")
    if raw == "null":
        return Scalar("null", None, "null")
    if _INT_RE.match(raw):
        # Normalize: -0 -> 0, no leading zeros (already enforced by regex)
        v = raw
        if v == "-0":
            v = "0"
        return Scalar("integer", v, v)
    if _DEC_RE.match(raw):
        # Preserve exactly
        return Scalar("decimal", raw, raw)
    # Otherwise treat as atom
    if not ATOM_RE.match(raw):
        raise ParseError("L010_INVALID_ATOM", f"Invalid atom: {raw!r}", cur.line, cur.col)
    return Scalar("atom", raw, raw)


def parse_attrs_payload(s: str, start_line: int = 1) -> List[Tuple[str, Scalar]]:
    """Parse {key:value,key:value} payload. Input s includes surrounding braces."""
    cur = StringCursor(s, line=start_line)
    skip_inline_ws(cur)
    if cur.peek() != "{":
        raise ParseError("S006_INVALID_ATTRS", "Expected {", cur.line, cur.col)
    cur.next()
    pairs: List[Tuple[str, Scalar]] = []
    skip_inline_ws(cur)
    if cur.peek() == "}":
        cur.next()
        return pairs
    while True:
        skip_inline_ws(cur)
        # Read key
        kstart = cur.i
        while cur.peek() and cur.peek() not in " \t:,":
            if cur.peek() == "}":
                break
            cur.next()
        key = cur.s[kstart:cur.i]
        if not key:
            raise ParseError("L003_INVALID_KEY", "Empty key", cur.line, cur.col)
        skip_inline_ws(cur)
        if cur.peek() != ":":
            raise ParseError("S006_INVALID_ATTRS", "Expected : after key", cur.line, cur.col)
        cur.next()
        v = parse_scalar(cur)
        pairs.append((key, v))
        skip_inline_ws(cur)
        c = cur.peek()
        if c == ",":
            cur.next()
            skip_inline_ws(cur)
            if cur.peek() == "}":
                cur.next()
                break
            continue
        elif c == "}":
            cur.next()
            break
        else:
            raise ParseError("S006_INVALID_ATTRS", f"Expected , or }} got {c!r}", cur.line, cur.col)
    return pairs

File: codec-cortex-node/src/scalars.js
'use strict';

/** Scalar value model and lexer for CORTEX 0.1. */

function toNfc(value) {
  return String(value).normalize('NFC');
}

function utf8Bytes(value) {
  return Buffer.from(String(value), 'utf8');
}

function pythonRepr(value) {
  const s = String(value);
  const quote = s.includes("'") && !s.includes('"') ? '"' : "'";
  let out = '';
  for (const ch of s) {
    if (ch === '\\') out += '\\\\';
    else if (ch === quote) out += `\\${ch}`;
    else if (ch === '\n') out += '\\n';
    else if (ch === '\r') out += '\\r';
    else if (ch === '\t') out += '\\t';
    else if (ch === '\b') out += '\\x08';
    else if (ch === '\f') out += '\\x0c';
    else {
      const cp = ch.codePointAt(0);
      if (cp < 0x20 || cp === 0x7f) out += `\\x${cp.toString(16).padStart(2, '0')}`;
      else out += ch;
    }
  }
  return `${quote}${out}${quote}`;
}

class Scalar {
  constructor(kind, value, lexeme) {
    this.kind = kind;
    this.value = value;
    this.lexeme = lexeme;
  }

  clone() {
    if (this.kind === 'list') {
      return new Scalar('list', this.value.map((item) => item.clone()), this.lexeme);
    }
    return new Scalar(this.kind, this.value, this.lexeme);
  }
}

class ParseError extends Error {
  constructor(code, message, line = 0, col = 0) {
    super(`${code} @ ${line}:${col} — ${message}`);
    this.name = 'ParseError';
    this.code = code;
    this.message = message;
    this.line = line;
    this.col = col;
  }
}

const ATOM_BODY_SOURCE = '[_A-Za-z][_A-Za-z0-9./:@+%$-]*';
const ATOM_RE = new RegExp(`^(?:\\$[0-9]+:)?${ATOM_BODY_SOURCE}$`);
const INT_RE = /^-?(0|[1-9][0-9]*)$/;
const DEC_RE = /^-?(0|[1-9][0-9]*)\.[0-9]+$/;

function isAtomLexeme(value) {
  const s = String(value);
  if (!s) return false;
  if (/[\s\[\]{},"|]/u.test(s)) return false;
  if (Array.from(s).length > 32) return false;
  return ATOM_RE.test(s);
}

const ESCAPE_MAP = new Map([
  ['"', '"'],
  ['\\', '\\'],
  ['n', '\n'],
  ['r', '\r'],
  ['t', '\t'],
  ['b', '\b'],
  ['f', '\f'],
]);

const REV_ESCAPE = new Map([
  ['"', '\\"'],
  ['\\', '\\\\'],
  ['\n', '\\n'],
  ['\r', '\\r'],
  ['\t', '\\t'],
  ['\b', '\\b'],
  ['\f', '\\f'],
]);

function parseStringLiteral(source) {
  const chars = Array.from(String(source));
  const out = [];
  let i = 0;
  while (i < chars.length) {
    const c = chars[i];
    if (c === '\\') {
      if (i + 1 >= chars.length) {
        throw new ParseError('L005_INVALID_STRING', 'Trailing backslash in string');
      }
      const next = chars[i + 1];
      if (ESCAPE_MAP.has(next)) {
        out.push(ESCAPE_MAP.get(next));
        i += 2;
      } else if (next === 'u') {
        const hex = chars.slice(i + 2, i + 6).join('');
        if (hex.length !== 4 || !/^[0-9a-fA-F]{4}$/.test(hex)) {
          throw new ParseError('L005_INVALID_STRING', 'Bad \\u escape');
        }
        out.push(String.fromCharCode(Number.parseInt(hex, 16)));
        i += 6;
      } else if (next === '/') {
        out.push('/');
        i += 2;
      } else {
        throw new ParseError('L005_INVALID_STRING', `Unknown escape \\${next}`);
      }
    } else if (c === '"') {
      throw new ParseError('L005_INVALID_STRING', 'Unescaped quote in string body');
    } else {
      out.push(c);
      i += 1;
    }
  }
  return out.join('');
}

function emitStringLiteral(value) {
  const out = [];
  for (const ch of String(value)) {
    if (REV_ESCAPE.has(ch)) {
      out.push(REV_ESCAPE.get(ch));
      continue;
    }
    const codePoint = ch.codePointAt(0);
    if (codePoint < 0x20) {
      out.push(`\\u${codePoint.toString(16).toUpperCase().padStart(4, '0')}`);
    } else if (codePoint === 0x7f) {
      out.push('\\u007F');
    } else {
      out.push(ch);
    }
  }
  return `"${out.join('')}"`;
}

class StringCursor {
  constructor(source, line = 1, col = 1) {
    this.s = String(source);
    this.chars = Array.from(this.s);
    this.i = 0;
    this.line = line;
    this.col = col;
  }

  eof() {
    return this.i >= this.chars.length;
  }

  peek(offset = 0) {
    const index = this.i + offset;
    return index >= 0 && index < this.chars.length ? this.chars[index] : '';
  }

  next() {
    const c = this.chars[this.i];
    this.i += 1;
    if (c === '\n') {
      this.line += 1;
      this.col = 1;
    } else {
      this.col += 1;
    }
    return c;
  }

  slice(start, end = this.i) {
    return this.chars.slice(start, end).join('');
  }
}

function skipInlineWs(cursor) {
  while (cursor.peek() === ' ' || cursor.peek() === '\t') cursor.next();
}

function parseScalar(cursor, inList = false) { // eslint-disable-line no-unused-vars
  skipInlineWs(cursor);
  const c = cursor.peek();
  if (c === '"') return parseStringScalar(cursor);
  if (c === '[') return parseListScalar(cursor);
  return parseAtomOrNumber(cursor);
}

function parseStringScalar(cursor) {
  if (cursor.peek() !== '"') throw new Error('parseStringScalar requires an opening quote');
  cursor.next();
  const body = [];
  while (true) {
    const c = cursor.peek();
    if (c === '') {
      throw new ParseError('L005_INVALID_STRING', 'Unterminated string', cursor.line, cursor.col);
    }
    if (c === '"') {
      cursor.next();
      break;
    }
    if (c === '\\') {
      body.push(c);
      cursor.next();
      const next = cursor.peek();
      if (next === '') {
        throw new ParseError('L005_INVALID_STRING', 'Trailing backslash', cursor.line, cursor.col);
      }
      body.push(next);
      cursor.next();
      if (next === 'u') {
        for (let j = 0; j < 4; j += 1) {
          const h = cursor.peek();
          if (h === '') {
            throw new ParseError('L005_INVALID_STRING', 'Bad \\u escape', cursor.line, cursor.col);
          }
          body.push(h);
          cursor.next();
        }
      }
    } else {
      body.push(c);
      cursor.next();
    }
  }
  const raw = body.join('');
  const value = parseStringLiteral(raw);
  return new Scalar('string', value, emitStringLiteral(value));
}

function parseListScalar(cursor) {
  if (cursor.peek() !== '[') throw new Error('parseListScalar requires [');
  cursor.next();
  const items = [];
  skipInlineWs(cursor);
  if (cursor.peek() === ']') {
    cursor.next();
    return new Scalar('list', items, '[]');
  }
  while (true) {
    const value = parseScalar(cursor, true);
    items.push(value);
    skipInlineWs(cursor);
    const c = cursor.peek();
    if (c === ',') {
      cursor.next();
      skipInlineWs(cursor);
    } else if (c === ']') {
      cursor.next();
      break;
    } else {
      throw new ParseError('L007_INVALID_LIST', `Expected , or ] got ${pythonRepr(c)}`, cursor.line, cursor.col);
    }
  }
  return new Scalar('list', items, `[${items.map((item) => item.lexeme).join(',')}]`);
}

function parseAtomOrNumber(cursor) {
  const start = cursor.i;
  while (true) {
    const c = cursor.peek();
    if (c === '' || ' \t\r\n,}]|'.includes(c)) break;
    cursor.next();
  }
  const raw = cursor.slice(start, cursor.i);
  if (raw === 'true') return new Scalar('boolean', true, 'true');
  if (raw === 'false') return new Scalar('boolean', false, 'false');
  if (raw === 'null') return new Scalar('null', null, 'null');
  if (INT_RE.test(raw)) {
    const value = raw === '-0' ? '0' : raw;
    return new Scalar('integer', value, value);
  }
  if (DEC_RE.test(raw)) return new Scalar('decimal', raw, raw);
  if (!ATOM_RE.test(raw)) {
    throw new ParseError('L010_INVALID_ATOM', `Invalid atom: ${pythonRepr(raw)}`, cursor.line, cursor.col);
  }
  return new Scalar('atom', raw, raw);
}

function parseAttrsPayload(source, startLine = 1) {
  const cursor = new StringCursor(source, startLine, 1);
  skipInlineWs(cursor);
  if (cursor.peek() !== '{') {
    throw new ParseError('S006_INVALID_ATTRS', 'Expected {', cursor.line, cursor.col);
  }
  cursor.next();
  const pairs = [];
  skipInlineWs(cursor);
  if (cursor.peek() === '}') {
    cursor.next();
    return pairs;
  }
  while (true) {
    skipInlineWs(cursor);
    const keyStart = cursor.i;
    while (cursor.peek() && !' \t:,'.includes(cursor.peek())) {
      if (cursor.peek() === '}') break;
      cursor.next();
    }
    const key = cursor.slice(keyStart, cursor.i);
    if (!key) throw new ParseError('L003_INVALID_KEY', 'Empty key', cursor.line, cursor.col);
    skipInlineWs(cursor);
    if (cursor.peek() !== ':') {
      throw new ParseError('S006_INVALID_ATTRS', 'Expected : after key', cursor.line, cursor.col);
    }
    cursor.next();
    const value = parseScalar(cursor);
    pairs.push([key, value]);
    skipInlineWs(cursor);
    const c = cursor.peek();
    if (c === ',') {
      cursor.next();
      skipInlineWs(cursor);
      if (cursor.peek() === '}') {
        cursor.next();
        break;
      }
    } else if (c === '}') {
      cursor.next();
      break;
    } else {
      throw new ParseError('S006_INVALID_ATTRS', `Expected , or } got ${pythonRepr(c)}`, cursor.line, cursor.col);
    }
  }
  return pairs;
}

module.exports = {
  Scalar,
  ParseError,
  StringCursor,
  ATOM_BODY_SOURCE,
  ATOM_RE,
  INT_RE,
  DEC_RE,
  _INT_RE: INT_RE,
  _DEC_RE: DEC_RE,
  toNfc,
  to_nfc: toNfc,
  utf8Bytes,
  pythonRepr,
  python_repr: pythonRepr,
  utf8_bytes: utf8Bytes,
  isAtomLexeme,
  is_atom_lexeme: isAtomLexeme,
  parseStringLiteral,
  parse_string_literal: parseStringLiteral,
  emitStringLiteral,
  emit_string_literal: emitStringLiteral,
  parseScalar,
  parse_scalar: parseScalar,
  skipInlineWs,
  skip_inline_ws: skipInlineWs,
  parseStringScalar,
  parse_string_scalar: parseStringScalar,
  parseListScalar,
  parse_list_scalar: parseListScalar,
  parseAtomOrNumber,
  parse_atom_or_number: parseAtomOrNumber,
  parseAttrsPayload,
  parse_attrs_payload: parseAttrsPayload,
};
